### ISSUE TITLE
Review design of `hook`, `cmpt` and `state` functions

### ISSUE THREAD

**Issue by nasser** [arcadia-unity/Arcadia#197]
Some issues have arisen around the semantics of state update and hooks. Namely we want to explore:

* The ability to atomically update the whole state atom
* Making `hook+` and `cmpt+` *replace* rather than append

**Comment by selfsame** [comment]
the api design for `cmpt+` and `ensure-cmpt` works well, as the user will frequently need multiple component types on an object.

This makes less sense for hooks, as functions are stateless.  I think the two design change options would be:

* have `hook+` replace based on it's key ident (the var/fn hash when not explicitly provided)
* leave `hook+` as is, and add an `ensure-hook`.  One caveat there is whether `ensure-hook` replaces or leaves the original (following the `ensure-cmpt` design)

**Comment by selfsame** [comment]
some thoughts on the `state` api:

There's good rationale for the `kw` arg, and not having the state functions just use `swap!` and `reset!` on the `.state` atom.  I think the feature missing for the user is being able to set the initial state of an object in a single go. 

Extending `set-state!` could allow that:
```clj
(defn set-state!
  "Updates the state of object `go` with funciton `f`."
  ([go v]
    (let [c (ensure-state go)]
     (reset! (.state c) v)))
  ([go kw v]
   (let [c (ensure-state go)]
     (swap! (.state c) assoc kw v))))
```

`update-state!` can't be extended as nicely (could check `keyword?` on the second arg, but that's not idiomatic)

**Comment by timsgardner** [comment]
I wonder if we really need keyless hooks. The idea of putting something in I can't cleanly remove or even access skeezes me out.

I dislike the notion of keying things to the hash of their function if no key is provided (which is to say, by _default_). This would make our behavior depend on the semantics of objects with reference identity (`fn`s), and gum up live programming much more than the drudgery of typing out some key. 

If we really need to support unkeyed hooks, let's say each hook type (`:update`, `:start`, etc) gets at most one free hook `fn`, which can always be accessed by asking for the hook by the hook type alone, and which replaces any unkeyed hooks that came before it. This way we can retain near-idempotence, and avoid making live programming a probable memory and behavior leak.

I think we should definitely have a `replace-hook` and an `ensure-hook`... or just a `put-hook` that knocks out any and all predecessors (on a key!), replacing them with the new hook. I'd prefer `put-hook` to `replace-hook` because it's shorter to type and it doesn't imply that another hook is necessarily already there. We're presumably not going to be making mutable hooks, so there's a lot less pressure for something like `ensure-hook`, and we can just cleanly replace a hook with something else every time. This is much closer to how maps work.

Do we need to accumulate multiple hook functions on a single key? Again, it kills idempotence, and introduces a lot of semantic complexity, with particularly dire consequences for live programming. The default behavior is to pile up hooks. Since the expected behavior of these hooks is to mutate something, the overall behavior will probably change (and in an order-dependent way!) every time we add a new hook. The only clean way to program in this context is to defensively wipe out all preexisting hook functions on a given key, and then add a new one. It would be better to just have one hook function per hook type and key to begin with. It would also mean we can be confident of retrieving a single hook function or nil when we look it up by key, rather than usually-a-hook-function-but-sometimes-a-collection-of-hook-functions, or the-first-hook-function-out-of-possibly-many.

We should get much clearer on what we're returning from these functions. Sometimes returning the object we're hooking, stating, or cmpt-ing, and sometimes returning the thing we're sticking into that object, seems flakey and confusing. With `cmpt+` or `ensure-cmpt` we definitely need to return the actual component, because there's no other way of getting it (and `ensure-cmpt` is a combined getter/setter/constructor anyway). The pattern of returning the object in these cases feels to me like an OOP holdover made obsolete by macros like `doto`. It's also more consistent with the Clojure api: `set!` returns the new value, not the object holding the field.

**Comment by jeaye** [comment]
I agree that requiring the hook key is a good default. From there, I would strongly prefer that we use a more Clojure-like naming scheme and opt for something like `assoc-hook`. Given that hooks would be a key/value pair, `assoc` is exactly what we'd be doing. Furthermore, `dissoc-hook` follows suit.

**Comment by timsgardner** [comment]
My reservation about `assoc-`, `dissoc-`, etc, is that the corresponding Clojure functions seem essentially stateless, whereas what we're doing here is very imperative. `ensure-cmpt` is closer to `update` than `get` or `assoc` in some ways, for example, since one generally uses it just before mutating the retrieved component. And if we do decide to keep accumulating rather than replacing `hook`s , `assoc` is even less appropriate, since we're doing something closer to `(update-in [hook-type k] conj hook-fn)`.

**Comment by timsgardner** [comment]
We should definitely have a way to update the state atom as a whole. Without this we have no thread-safe way of updating the state atom that takes multiple key-value pairs in it into account.

**Comment by timsgardner** [comment]
Originally we wanted to protect the user against stepping on someone else's toes in the state atom, but if we take that beyond namespace-qualified keys we quickly get into OOP visibility nonsense, and lose clean concurrency. I don't think it's worth it anymore. It's easy enough to write libraries that mind their own business in a map.

**Comment by timsgardner** [comment]
(warming to `assoc-hook`, `dissoc-hook`)

**Comment by timsgardner** [comment]
If we did go with `assoc-hook` etc we should probably return the object itself, just to be consistent with the feel of the corresponding clojure functions. Is that worth it?

**Comment by jeaye** [comment]
Yes! It also allows for more idiomatic Clojure, like the threading macros.

```clojure
(-> (assoc-hook obj :start :foo #'foo)
    (assoc-hook :update :bar #'bar))
```

**Comment by jeaye** [comment]
To take things further, we could have `conj-cmpt` and `disj-cmpt`, since they aren't key/value pairs. That could also mean `conj-child`, `disj-child`, etc.

**Comment by timsgardner** [comment]
Again, not sure it is more idiomatic for Clojure when mutation's involved:

```clojure
(doto obj
  (put-hook :start ::foo #'foo)
  (put-hook :update :bar #'bar))
```
Also, when it comes to `state`, odds of wanting to refer back to to the value itself seem greater:
```clojure
(when-let [new-state (put-state obj ::stuff (calculate-state obj))]
  (println "adding new state: " new-state))
```
seems nicer than
```clojure
(if-let [new-state (calculate-state obj)]
  (do (put-state obj ::stuff new-state)
      (println "adding new state: " new-state))
  (put-state obj ::stuff nil))
```
...again, on the analogy of `set!`

**Comment by nasser** [comment]
I don't know about *requiring* a keyword for hooks. I agree that leaning on equality semantics of functions is weird, and keywords are important for anonymous functions, but for vars this starts to look like

```clojure
(hook+ gobj :update :move-player #'move-player)
```

which is tedious. I can't think what that keyword would be aside from the name of the var. Also, the equality semantics of Vars are well defined. 

```clojure
user> (defn foo [x] x)
#'user/foo
user> (= #'foo #'foo)
true
user> (identical? #'foo #'foo)
true
```

Which is why the keywordless signatures still make sense to me, in addition to the keyworded ones. I suppose Vars are the only argument in this case, but they're an important one as hooking Vars is the only way to serialize, and they play very nicely with live coding.

In this scheme, it remains possible to add a function without a keyword and have a hard time getting it out, which is annoying. We could either:

1. In the keywordless signature, check if the fn is a var and throw an exception/issue a warning otherwise
2. Do nothing, and document the behavior well

**Comment by nasser** [comment]
I like the idea of getting closer to Clojure names and semantics. Hooks can be thought of as sorted sets, so maybe `assoc-hook` and `dissoc-hook` make sense. `conj-hook` could work for the above mentioned keywordless hooks for vars, which could serve to preserve the ergonomics of vars while preventing unkeyed functions otherwise.

```clojure
(assoc-hook gobj :rotate #(.. % transform (Rotate 1 0 0)))
(conj-hook gobj #'move-player)
(dissoc-hook gobj :rotate)
(dissoc-hook gobj #'move-player)
```

Under this scheme there are no overloads. `assoc-hook` always takes a key, and `conj-hook` never does.

**Comment by jeaye** [comment]
I think, if you want to allow both assoc and conj, you should allow both dissoc and disj. I'm glad you're into moving the naming closer to Clojure.

**Comment by timsgardner** [comment]
another thing: I think maybe `with-cmpt` should use `ensure-cmpt` rather than `cmpt`. Seems to flow better.

**Comment by timsgardner** [comment]
Regarding hooks, I think we might be talking about two functions. Names are important but they're also distracting, so I'm going to call the two functions `zum` and `sko`.

One very important property of keyed collections in a dynamic context is that the keys are generally more stable than their values. Given a key-value pair in a map, it is more common for the value to be swapped out than for the key to be changed. We look things up by keys, so keys are the question, and the value is the answer. We express differences in data with the "same structure" by giving different values to the same keys -- different answers to the same question. 

If I key a var to itself, I can remove it, but only if I remember that I put it there in the first place. Otherwise it just piles up alongside any other vars I've keyed to themselves. I lose the very important ability to change the meaning while retaining the structure. It's better than keying to really transient things like Ifn hashes (or randomly-generated ids, as in Javascript), but it isn't _much_ better.

I think we should have a function (`zum`, for now) that is all about this kind of predictable, keyed access. The meaning and implications of its call sites will be clear:
```clojure
(zum obj :update ::bounce #'bounce-1)
```
will always mean the `::bounce` functionality is being dealt with by `#'bounce-1`. If later on we change our minds and want `#'bounce-2` instead, but don't want to redefine `#'bounce-1` (which other things might be using, or we might not control, or whatever), we can just say:
```clojure
(zum obj :update ::bounce #'bounce-2)
```
In this case we _know_ what this object's implementation of `::bounce` is, and don't have to worry about the highly order-dependent, imperative, difficult question of what it _was_. We know exactly what we're going to be dealing with, `::bounce`-wise, after evaluating this expression.

Names are sometimes annoying though, and if we don't want to have to concoct one, I still think we have a good option. Rather than keying the hook value to itself (even as a var), let's key it to the hook type. This way `zum` retains its stability and declarativity.
```clojure
(zum obj :update #'bounce-1)
```
then later:
```clojure
(zum obj :update #'bounce-2)
```
Again, the latter expression would knock out the former, and we know just where we stand afterwards. The result of this 3-argument version would be the same as `(zum obj :update :update #'bounce-1)`.

But maybe we also want to support another kind of programming, that places additional emphasis on vars; vars can be redefined, after all, so maybe there's less need to keep them rigorously keyed, and maybe we don't want the pressure of having to invent new names (keys) for things we've already effectively named by defining the var. Let's call the var-centric function `sko`. 

`sko` always takes 3 arguments: a GameObject, a hook type, and a var. `sko` works like the 3-argument version of `zum`, but keys the var to itself, rather than to the update type. Where `(zum obj :update #'bounce-1)` is the same as  `(zum obj :update :update #'bounce-1)`, `(sko obj :update #'bounce-1)` is the same as `(zum obj :update #'bounce-1 #'bounce-1)`. `sko` is all about vars, so it throws an exception if it's given something else for its third argument.

I think we should have `zum` and `sko` (with different names).

**Comment by selfsame** [comment]
I worry this is starting to get over engineered and would be steering users into conceptualizing the hook components as stores of state that are meant to be continually reconfigured.

A more idiomatic pattern for the previous example would be:
```
(defn frog-update [o]
  (if-let [f (state o ::bounce)]
    (f)))

(hook+ frog :update #'frog-update)

(set-state! frog ::bounce #'bounce-2)
```

I still think `hook-` via keys is a responsible part of the API though, so would propose:

* leave the optional key arity as is, where the function value is used as the key when none is supplied
* `hook+` replaces by key
* documentation makes clear that "hooks use vars", and that fn literals/refs break serialization and will need user keys to be removed.

Re: the semantics, I like Arcadia's `hook+` | `cmpt-` style.  Using `conj` `assoc` style verbs would seem to indicate that components & hooks are satisfying a variant of clojure.core protocols.

**Comment by selfsame** [comment]
`ArcadiaBehaviour`'s `fnIndexes` map is not serialized right now, meaning you can only `hook-` functions that were added after deserialization.  This could be an issue, or an argument for simplifying things by ditching keys & `fnIndexes` and having `hook-` remove fns via their hash.

**Comment by nasser** [comment]
Implemented by @timsgardner's `hook-help` branch (61df705473ceddba0bf70c1d51cc3b6f22eb1be8 - 5d1ba134e1ac384569348593f79d3a1223826361)

**Comment by pjago** [comment]
How about having hook and cmpt return nil if their obj is nil? As in `(get nil :football)` returning nil in clojure core. Then we can do something like `(if-let [mr (cmpt (parent gobj) MeshRenderer)] ...)` work.
